nncf/common/hook_handle.py

```diff
@@ -0,0 +1,82 @@
+# Copyright (c) 2023 Intel Corporation
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#      http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import weakref
+from abc import ABC
+from abc import abstractmethod
+from typing import Any, Dict, Union
+
+
+class HookHandleInterface(ABC):
+    """
+    A handle to remove a hook.
+    """
+
+    @property
+    @abstractmethod
+    def hook_id(self) -> str:
+        """
+        Key to use to retrieve handle from the registry.
+        """
+
+    @abstractmethod
+    def remove(self) -> None:
+        """
+        Removes added operation from registered hooks registry if it is possible.
+        """
+
+
+class _HookHandle(HookHandleInterface):
+    """
+    Privat hook handle implementation which should be instanciated only by
+    the add_op_to_registry function.
+    """
+
+    def __init__(self, hooks_registry: Dict[Any, Any], hook_id: str):
+        """
+        :param hooks_registry: A dictionary of hooks, indexed by hook `id`.
+        """
+        self.hooks_registry_ref = weakref.ref(hooks_registry)
+        self._hook_id = hook_id
+
+    @property
+    def hook_id(self) -> Union[int, str]:
+        return self._hook_id
+
+    def remove(self) -> None:
+        """
+        Removes added operation from registered hooks registry if it is possible.
+        """
+        if self.hooks_registry_ref is None:
+            return
+
+        hooks_registry = self.hooks_registry_ref()
+        if hooks_registry is not None and self.hook_id in hooks_registry:
+            del hooks_registry[self.hook_id]
+            self.hooks_registry_ref = None
+
+
+def add_op_to_registry(hooks_registry: Dict[Any, Any], op: Any) -> HookHandleInterface:
+    """
+    Registers op into the hooks_registry and returns HookHandler instance.
+
+    :param hooks_registry: A dictionary of hooks, indexed by hook `id`.
+    :param op: Operation to set to registered hooks registry.
+    :return: HookHandle instance binded with hook registry and given op that
+        able to remove op from the hooks_registry.
+    """
+    if hooks_registry:
```

Comment:
What about encapsulating registry and requests for adding and removing op in the single `HookRegistry` class?

Reply:
Such approach was rejected, please see https://github.com/openvinotoolkit/nncf/pull/2313#discussion_r1449966113